utils/cautils/acmeutils.go

```diff
@@ -461,10 +481,14 @@ func (af *acmeFlow) GetCertificate() ([]*x509.Certificate, error) {
 		}
 
 		_csr := &x509.CertificateRequest{
-			Subject: pkix.Name{
+			DNSNames:    dnsNames,
+			IPAddresses: ips,
+		}
+		subjectIsNotAnIP := net.ParseIP(af.subject) == nil
+		if subjectIsNotAnIP {
+			_csr.Subject = pkix.Name{
```

Comment:
Is this in the spec? Why subject cannot be an IP?

Reply:
@maraino: It is not; it would've been a lot simpler if it were 😅 

So far I haven't found a normative reference that explicitly forbids putting IPs in the Subject Common Name. I did find and read some general (best) practices and conclusions (that others reached after reading the non-conclusive RFCs), though. These were also partially based on practices for the Web PKI (i.e. CAB Forum). The general consensus was that the Common Name should actually not be used anymore, as it's deprecated by Subject Alternative Names and many clients operate on just those. Next to that, putting an IP in the Subject seems to be discouraged in general. It is used when having to support legacy software, though.

Technically I don't see a reason to put anything in the Subject Common Name at all. It has has been deprecated, and clients should've been updated to not use the Common Name when performing their checks. Changing that logic could break some current deployments, so the decision to not fill it shouldn't be taken too lightly, IMO, so it's good that you and @dopey are critical about this. I also understand that there are legacy use cases that need to be supported, but I think those should be exceptions and not the default.

Apart from the above, I did in fact try to set an IP in the Subject before pushing my changes. IIRC I observed that step-ca would perform two challenges for a single IP identifier if it was also in the Subject _or_ would report a mismatch between the CSR and the ACME Order object, depending on my approach. I'm pretty sure these can be fixed if it is decided to support IPs as the Subject, but it might involve a bit more logic and/or changes in step-ca also. I'd probably want to refactor some of this logic to make it more easily testable in that case, then.

Only partially related, but it may influence the decision: yesterday I had a chat with @tashian about the CLI being able to renew certificates when using the ACME protocol. Currently renewals work, but only after generating a new account key and new authorizations are performed. The CLI is effectively stateless in this regard. It may be good if the CLI ACME implementation would be changed to reuse an existing account key, especially if we also want to support ACME EAB. Now that's a larger effort, but we may be able to integrate [CertMagic](https://github.com/caddyserver/certmagic) or [acmez](https://github.com/mholt/acmez) instead, to not have to (re)do all of this ourselves. Both do not seem to provide a way to add anything as the Subject Common Name, though.

Reply:
I have to check what the CAB Forum exactly says, but my understanding is that it should not be used to authenticate a server, but I think it's ok to have the CN as an information field for a server.

Reply:
I'm ok with integrating an acme client library, ours just does the job, but for example, it does more requests than necessary as it doesn't reuse the nonces. But at the same time, I don't think we need all features that `lego` or `certbot` have. 

We basically have to integrate it well with our `step-ca`, but at the same time, you should be able to do basic stuff with let's encrypt or any other standard ACME server.

Reply:
> I have to check what the CAB Forum exactly says, but my understanding is that it should not be used to authenticate a server, but I think it's ok to have the CN as an information field for a server.

That makes sense, indeed. We can treat it as an opaque string if provided as an argument or set in a CSR. The default could become to not set it; adding a flag to set it seems appropriate to me. The ACME authorizations shouldn't take the subject into account, which might need a change in step-ca. I'll update this PR with the changes soon.

> I'm ok with integrating an acme client library, ours just does the job, but for example, it does more requests than necessary as it doesn't reuse the nonces. But at the same time, I don't think we need all features that lego or certbot have.
>
>We basically have to integrate it well with our step-ca, but at the same time, you should be able to do basic stuff with let's encrypt or any other standard ACME server.


I'll have a look at what the impact is for changing the ACME client implementation, but won't start implementing one till we decide to do so.

Reply:
If we decide to use a library it should be part of a separate PR.